build/parseSpec.c

```diff
@@ -1060,6 +1053,67 @@ static rpmSpec parseSpec(const char *specFile, rpmSpecFlags flags,
 	}
     }
 
+    /* Check for description in each package */
+    for (Package pkg = spec->packages; pkg != NULL; pkg = pkg->next) {
+	if (!headerIsEntry(pkg->header, RPMTAG_DESCRIPTION)) {
+	    rpmlog(RPMLOG_ERR, _("Package has no %%description: %s\n"),
+		   headerGetString(pkg->header, RPMTAG_NAME));
+	    goto errxit;
+	}
+    }
+
+    /* Add arch, os and platform, self-provides etc for each package */
+    addTargets(spec->packages);
+
+    /* Check for encoding in each package unless disabled */
+    if (!(spec->flags & RPMSPEC_NOUTF8)) {
+	int badenc = 0;
+	for (Package pkg = spec->packages; pkg != NULL; pkg = pkg->next) {
+	    if (checkForEncoding(pkg->header, 0) != RPMRC_OK) {
+		badenc = 1;
+	    }
+	}
+	if (badenc)
+	    goto errxit;
+    }
+
+    closeSpec(spec);
+
+exit:
+    return spec;
+
+errxit:
+    if (!secondary)
+	rpmSpecFree(spec);
```

Comment:
Indeed. The way it gets called, it probably doesn't matter as parseSpecSection() returns an explicit NULL on error, but freeing the struct is the responsibility of the caller.

There's also something broken in the error handling: an invalid spec part does not abort the build, although an error message can be seen in the log. Do add a test for that too, just reuse the same spec but leave a mandatory part out or something.

Reply:
I think just moving the initial spec allocation (+ other setup) to parseSpecSection() should make it a whole lot less lopsided wrt the allocation, and thus less ugly. We can't help the recursing from buildarch, but we're not going to make things *worse*.

Reply:
Another possibility may be a kind of opposite direction: handle it inline specParse(), just move stuff that doesn't belong there out of the way. I have no idea why I planted the openmp thread fubar right in the middle of specParse(), it sure does not have to be just there, it may just as well be done from rpmSpecParse() instead, probably ditto for various other pieces.

Reply:
Of course it needs a spec pointer. The point here is to avoid two paths of error cleanup in different functions which is just the kind of murky environment that all manner of unpleasant creatures thrive. We can't fix everything that is crazy in there, but never ever make bad code worse. Find a way, I'm sure there is one or more. My musings above were just ways I'd look into resolving it, there may be others/better ones.

Reply:
OK, I was able to at least get rid of the `rpmSpecFree` in `parseSpec` which is just not needed. The API sets some hard limits on what we can do here. While parseSpec is returning NULL on error we can't free the spec when parsing dynamic spec parts as the surrounding code does not expect the spec object to go away at that point.

Reply:
I am going to try another variant with the spec being a & parameter and the freeing is done outside of parseSpecSection(). That may be more to the book even if it doesn't really simplify things.

Reply:
Hmm, the API shouldn't get much in the way as parseSpec() and friends is all internal affairs.
I'm too cross-eyed to look at this today but glad you found a way forward,